src/System.Linq/src/System/Linq/Concat.cs

```diff
@@ -61,16 +60,15 @@ internal Concat2Iterator(IEnumerable<TSource> first, IEnumerable<TSource> second
 
             internal override ConcatIterator<TSource> Concat(IEnumerable<TSource> next)
             {
-                bool hasOnlyCollections = _first is ICollection<TSource> &&
-                                          _second is ICollection<TSource> &&
-                                          next is ICollection<TSource>;
+                bool hasOnlyCollections = next is ICollection<TSource> &&
+                                          _first is ICollection<TSource> &&
+                                          _second is ICollection<TSource>;
                 return new ConcatNIterator<TSource>(this, next, 2, hasOnlyCollections);
             }
 
             public override int GetCount(bool onlyIfCheap)
             {
-                int firstCount, secondCount;
-                if (!EnumerableHelpers.TryGetCount(_first, out firstCount))
+                if (!EnumerableHelpers.TryGetCount(_first, out int firstCount))
```

Comment:
I personally find out declarations like this difficult.  It's one thing if the declared variable is only used in the scope of the body of the if, or if the method is small, but otherwise I find it makes it difficult when later reading the method to figure out where the variable came from... I can no longer just look quickly up the left-hand side of the code and at the method signature; I instead need to essentially scan the full contents of many lines looking to see where it was defined.  I'd prefer we not use them in situations like this.